File: src/adapters/claude_code/extract_transcript.ts
import type { AssistantLine, UserLine } from './schemas.js';
import type { SegmentRole, RawSegment } from '../../schema/types.js';

// Re-export so existing imports from this module keep working. New code
// should import RawSegment directly from schema/types.js.
export type { RawSegment };

// Per-segment cap. Most assistant text fits well under 16 KB; tool_result
// content can be much larger (a `Read` of a file, a `Bash` of a build
// log, etc.). Indexing the entire body bloats the FTS5 index for low
// search value, so we truncate. The cap is a soft byte limit applied
// after the unicode-safe trimming used elsewhere.
const SEGMENT_CAP_BYTES = 16 * 1024;

export function extractTranscriptSegments(
  assistantLines: AssistantLine[],
  userLines: UserLine[],
): RawSegment[] {
  const out: RawSegment[] = [];

  for (const line of assistantLines) {
    let blockIdx = 0;
    for (const block of line.message.content) {
      if (block.type === 'text') {
        const text = (block as any).text;
        if (typeof text === 'string' && text.trim() !== '') {
          out.push({
            uid_suffix: `${line.uuid}:${blockIdx}`,
            timestamp: line.timestamp,
            role: 'assistant',
            text: capText(text),
          });
        }
      } else if (block.type === 'thinking') {
        const text = (block as any).thinking;
        if (typeof text === 'string' && text.trim() !== '') {
          out.push({
            uid_suffix: `${line.uuid}:${blockIdx}`,
            timestamp: line.timestamp,
            role: 'thinking',
            text: capText(text),
          });
        }
      }
      // tool_use blocks are not indexed here — their inputs are JSON
      // (commands, file paths) handled by the tool_calls table.
      blockIdx++;
    }
  }

  for (const line of userLines) {
    const c = line.message.content;
    if (typeof c === 'string') {
      if (c.trim() !== '') {
        out.push({
          uid_suffix: `${line.uuid}:0`,
          timestamp: line.timestamp,
          role: 'user',
          text: capText(c),
        });
      }
      continue;
    }
    if (!Array.isArray(c)) continue;
    let blockIdx = 0;
    for (const block of c as any[]) {
      if (!block) { blockIdx++; continue; }
      if (block.type === 'text' && typeof block.text === 'string' && block.text.trim() !== '') {
        out.push({
          uid_suffix: `${line.uuid}:${blockIdx}`,
          timestamp: line.timestamp,
          role: 'user',
          text: capText(block.text),
        });
      } else if (block.type === 'tool_result') {
        // tool_result.content can be a string or an array of {type:'text',text}
        // sub-blocks; we flatten both into one segment per result so a single
        // command's output is one searchable unit.
        const content = block.content;
        let combined = '';
        if (typeof content === 'string') {
          combined = content;
        } else if (Array.isArray(content)) {
          combined = content
            .map((b: any) => (typeof b?.text === 'string' ? b.text : ''))
            .filter(Boolean)
            .join('\n');
        }
        if (combined.trim() !== '') {
          out.push({
            uid_suffix: `${line.uuid}:${blockIdx}`,
            timestamp: line.timestamp,
            role: 'tool_result',
            text: capText(combined),
          });
        }
      }
      blockIdx++;
    }
  }

  return out;
}

// Cap a string to SEGMENT_CAP_BYTES of utf-8 with an ellipsis suffix when
// truncated. Walks back to a unicode boundary so we never split a multi-
// byte character in the middle.
function capText(s: string): string {
  if (Buffer.byteLength(s, 'utf8') <= SEGMENT_CAP_BYTES) return s;
  let cap = SEGMENT_CAP_BYTES;
  while (cap > 0 && Buffer.byteLength(s.slice(0, cap), 'utf8') > SEGMENT_CAP_BYTES - 1) {
    cap--;
  }
  return s.slice(0, cap) + '…';
}

File: python/argus/adapters/claude_code/extract_transcript.py
"""Extract searchable transcript segments from a session's lines."""
from __future__ import annotations

from ..base import RawSegment
from .schemas import AssistantLine, UserLine

# Per-segment byte cap. Tool_result content (a Read of a file, a Bash of a
# build log) can be enormous; indexing the entire body bloats FTS5 for
# very little search value.
SEGMENT_CAP_BYTES = 16 * 1024


def _sanitize(s: str) -> str:
    """Drop lone UTF-16 surrogates so the text can be encoded and stored.

    ``"\\ud800"`` is legal JSON, so ``json.loads`` hands back a str containing
    an unpaired surrogate — which has no UTF-8 encoding. Every later step
    (``str.encode`` here, and the SQLite insert) then raises
    ``UnicodeEncodeError``, and because that escaped the per-line handler it
    killed the entire ingest pass, not just the offending line.

    Replace rather than drop, so the corruption stays visible in search results
    instead of silently changing the text.
    """
    if s.isascii():  # fast path: the overwhelming majority of segments
        return s
    return s.encode("utf-8", errors="replace").decode("utf-8", errors="replace")


def _cap_text(s: str) -> str:
    """Truncate ``s`` to ``SEGMENT_CAP_BYTES`` UTF-8 bytes with an ellipsis."""
    s = _sanitize(s)
    encoded = s.encode("utf-8")
    if len(encoded) <= SEGMENT_CAP_BYTES:
        return s
    # Walk back to a unicode-safe boundary by decoding with errors='ignore'.
    truncated = encoded[:SEGMENT_CAP_BYTES].decode("utf-8", errors="ignore")
    return truncated + "…"


def extract_transcript_segments(
    assistant_lines: list[AssistantLine],
    user_lines: list[UserLine],
) -> list[RawSegment]:
    out: list[RawSegment] = []

    for line in assistant_lines:
        block_idx = 0
        for block in line.message.content:
            btype = block.get("type")
            if btype == "text":
                text = block.get("text")
                if isinstance(text, str) and text.strip():
                    out.append(
                        RawSegment(
                            uid_suffix=f"{line.uuid}:{block_idx}",
                            timestamp=line.timestamp,
                            role="assistant",
                            text=_cap_text(text),
                        )
                    )
            elif btype == "thinking":
                text = block.get("thinking")
                if isinstance(text, str) and text.strip():
                    out.append(
                        RawSegment(
                            uid_suffix=f"{line.uuid}:{block_idx}",
                            timestamp=line.timestamp,
                            role="thinking",
                            text=_cap_text(text),
                        )
                    )
            block_idx += 1

    for line in user_lines:
        c = line.message.content
        if isinstance(c, str):
            if c.strip():
                out.append(
                    RawSegment(
                        uid_suffix=f"{line.uuid}:0",
                        timestamp=line.timestamp,
                        role="user",
                        text=_cap_text(c),
                    )
                )
            continue
        if not isinstance(c, list):
            continue
        block_idx = 0
        for block in c:
            if not isinstance(block, dict):
                block_idx += 1
                continue
            btype = block.get("type")
            if btype == "text" and isinstance(block.get("text"), str) and block["text"].strip():
                out.append(
                    RawSegment(
                        uid_suffix=f"{line.uuid}:{block_idx}",
                        timestamp=line.timestamp,
                        role="user",
                        text=_cap_text(block["text"]),
                    )
                )
            elif btype == "tool_result":
                content = block.get("content")
                combined = ""
                if isinstance(content, str):
                    combined = content
                elif isinstance(content, list):
                    parts = []
                    for sub in content:
                        if isinstance(sub, dict):
                            t = sub.get("text")
                            if isinstance(t, str):
                                parts.append(t)
                    combined = "\n".join(p for p in parts if p)
                tu_id = block.get("tool_use_id")
                if combined.strip():
                    out.append(
                        RawSegment(
                            uid_suffix=f"{line.uuid}:{block_idx}",
                            timestamp=line.timestamp,
                            role="tool_result",
                            text=_cap_text(combined),
                            tool_use_id=tu_id if isinstance(tu_id, str) else None,
                        )
                    )
            block_idx += 1

    return out
